ReactiveCocoa/Swift/Flatten.swift

```diff
@@ -0,0 +1,571 @@
+	/// Flattens the inner producers sent upon `signal` (into a single signal of
+	/// values), according to the semantics of the given strategy.
+	///
+	/// If `signal` or an active inner producer fails, the returned signal will
+	/// forward that failure immediately.
+	///
+	/// `Interrupted` events on inner producers will be treated like `Completed`
+	/// events on inner producers.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func flatten(strategy: FlattenStrategy) -> SignalProducer<Value.Value, Error> {
+		switch strategy {
+		case .Merge:
+			return self.merge()
+
+		case .Concat:
+			return self.concat()
+
+		case .Latest:
+			return self.switchToLatest()
+		}
+	}
+}
+
+extension SignalType where Value: SignalType, Error == Value.Error {
+	/// Flattens the inner signals sent upon `signal` (into a single signal of
+	/// values), according to the semantics of the given strategy.
+	///
+	/// If `signal` or an active inner signal emits an error, the returned
+	/// signal will forward that error immediately.
+	///
+	/// `Interrupted` events on inner signals will be treated like `Completed`
+	/// events on inner signals.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func flatten(strategy: FlattenStrategy) -> Signal<Value.Value, Error> {
+		return self.map(SignalProducer.init).flatten(strategy)
+	}
+}
+
+extension SignalProducerType where Value: SignalType, Error == Value.Error {
+	/// Flattens the inner signals sent upon `signal` (into a single signal of
+	/// values), according to the semantics of the given strategy.
+	///
+	/// If `signal` or an active inner signal emits an error, the returned
+	/// signal will forward that error immediately.
+	///
+	/// `Interrupted` events on inner signals will be treated like `Completed`
+	/// events on inner signals.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func flatten(strategy: FlattenStrategy) -> SignalProducer<Value.Value, Error> {
+		return self.map(SignalProducer.init).flatten(strategy)
+	}
+}
+
+
+extension SignalType where Value: SignalProducerType, Error == Value.Error {
+	/// Returns a signal which sends all the values from producer signal emitted from
+	/// `signal`, waiting until each inner producer completes before beginning to
+	/// send the values from the next inner producer.
+	///
+	/// If any of the inner producers fail, the returned signal will forward
+	/// that failure immediately
+	///
+	/// The returned signal completes only when `signal` and all producers
+	/// emitted from `signal` complete.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	private func concat() -> Signal<Value.Value, Error> {
+		return Signal<Value.Value, Error> { observer in
+			self.observeConcat(ConcatState(observer: observer, disposable: nil))
+		}
+	}
+
+	private func observeConcat(state: ConcatState<Value.Value, Error>) -> Disposable? {
```

Comment:
If I were to write this, I would construct the `ConcatState` inside `observeConcat`—let the caller pass in the observer and an optional disposable. This more closely matches `Signal.observe()` and narrows the usage of `ConcatState` to inside `observeConcat`. (AFAICT it's an implementation detail—you don't need to extract those bits to a separate type.)